streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -847,7 +850,7 @@ class ZStream[-R, +E, +A](val process: ZManaged[R, E, Pull[R, E, A]]) extends Se
    * elements of this stream, and then emit the elements of the `other` stream.
    */
   final def concat[R1 <: R, E1 >: E, A1 >: A](other: => ZStream[R1, E1, A1]): ZStream[R1, E1, A1] =
```

Comment:
What exactly ended up being the issue with the implementation of concat based on ZManaged#switchable that you posted in discord? I though it was a bit nicer than introducing structure

Reply:
Good question and good to have it written in this PR for posterity.

The previous `++` would do something like this (written minimally on purpose):
```scala
def ++(other: => ZStream) = 
  ZStream {
    for {
      switched <- Ref.make(false).toManaged_
      currPull <- Ref.make(...)
      pull = {
        // get the current pull from the Ref, and pull it
        // if it ended and !switched, switch over to `other`
        // otherwise end
      }
    } yield pull
  }
```

As you can see, it allocates a bunch of resources that are retained for the existence of the stream `(self ++ other)`.

Now, let's consider what happens in `self.forever` (defined as `forever = self ++ forever`) - or any stream that concatenates recursively. When we evaluate `++`, it allocates these refs and starts pulling `self`. Once `self` is done, the ZManaged for `other` is opened and that one starts being pulled.

But, `other` is actually another `++`, so more `Ref` instances are allocated! This happens on every switch to the RHS, so every recursion allocates more `Ref`. That would be fine if we could somehow stop retaining the current `Ref`, but we can't "exit" the scope of the current ZManaged. This is the root of the problem. Once we're inside a scope, we can't exit it.

So this formulation is heap-unsafe. I tried pretty hard, but I couldn't work around it properly. I think that the way to solve it is to interpret the recursive structure "from outside", before entering the scope that is recursive.

I hope this makes sense!

Reply:
Ah, I think I understand. But this is a very subtle issue O.o

Thanks for the explanation and nice fix!